ffmpeg cleanup

ulmus-scott opened the pull request:
This now contains multiple pull requests with the following changes:

I decided to drop the MythTV changes to hide the use of FFmpeg's internal header libavformat/url.h since it only includes public headers and it is more involved to replace its use and I have not fully investigated it.

Otherwise, the MythTV changes I have kept are in:
- [ ] https://github.com/MythTV/mythtv/pull/568 stylistic change
- [ ] https://github.com/MythTV/mythtv/pull/565 lavc/utils-mythtv.c
- [ ] https://github.com/MythTV/mythtv/pull/576 Miscellaneous MythTV changes
- [ ] https://github.com/MythTV/mythtv/pull/557 ByteReader
- [ ] https://github.com/MythTV/mythtv/pull/577 BitReader 

I have rebased this onto master and made the following changes:

- renamed ff_codec_type_string to AVMediaTypeToString in mythtv/libs/libmythtv/decoders/avformatdecoder.cpp
- dropped changes concerning the use of FFmpeg's internal header libavformat/url.h since it only includes public headers and it is more involved to replace its use and I have not fully investigated it.
- only change to ffmpeg compared to the previous version is:
```
git diff ffmpeg-cleanup-backup -- mythtv/external/FFmpeg
diff --git a/mythtv/external/FFmpeg/libavcodec/libavcodec.v b/mythtv/external/FFmpeg/libavcodec/libavcodec.v
index 3f3dc96da0..f1d5e5bc77 100644
--- a/mythtv/external/FFmpeg/libavcodec/libavcodec.v
+++ b/mythtv/external/FFmpeg/libavcodec/libavcodec.v
@@ -1,11 +1,6 @@
 LIBAVCODEC_MAJOR {
     global:
         av*;
-        ff_ue_golomb_vlc_code;
-        ff_golomb_vlc_len;
-        ff_se_golomb_vlc_code;
-        ff_fft_init;
-        ff_fft_end;
     local:
         *;
 };
```

----
I am currently splitting this into multiple pull requests:

- [ ] https://github.com/MythTV/mythtv/pull/557
- [ ] bitreader (shared context with bytereader)
- [x] https://github.com/MythTV/mythtv/pull/558
- [x] https://github.com/MythTV/mythtv/pull/559
- [ ] https://github.com/MythTV/mythtv/pull/565
- [ ] https://github.com/MythTV/mythtv/pull/566
- [ ] ...
- [ ] remove now unnecessary modifications to FFmpeg all at once (shared context lines)

ulmus-scott commented:
Regarding libavcodec/get_bits.h (and golomb.h), these are private FFmpeg headers, so we would need to write our own (or use someone else's) C++ version.

Alternatively, we could use FFmpeg's parsers: [FFmpeg: Frame parsing](https://ffmpeg.org/doxygen/trunk/group__lavc__parsing.html)
Most of the information we use is stored in the AVCodecParserContext, but the frame rate is stored in the complementary AVCodecContext.

ulmus-scott commented:
I forgot to put this in the commit message for replace ff_codec_id_string() with avcodec_get_name():
The string is now lower case, instead of all capital letters; however, I don't anticipate that being an issue.

bennettpeter commented:
This has now been rebased into branch ffmpeg-resync.

ulmus-scott commented:
I rebased onto master and added the changes from https://github.com/ulmus-scott/FFmpeg/commits/rebase/4.4m2 .

However, some changes have been skipped, which I'll add later.

ulmus-scott commented:
The changes to FFmpeg should be equivalent to those in the devel/ffmpeg-resync branch, ignoring a few space changes.

This should allow bisecting the reversions if there are issues.

bennettpeter commented:
That streams file  looks like garbage. It should not be there. probably left over from when I was searching for occurrences of stream in the code. I will get rid of it.

ulmus-scott commented:
Do not merge.

I can confirm seeking in videos fails.  No crash, but exits playback.
```
E  AFD: av_seek_frame(ic, -1, 71615930311, 0) -- error
E  decoding error End of file (-541478725)
```

bennettpeter commented:
It looks like you found the problem Klaas was having.

I have not looked at all these commits. Not sure where these are going. I was planning on starting to incorporate ffmpeg's master into mythtv, using a branch from MythTV master. There are two ways to do it, either merge ffmpeg master into our copy of the ffmpeg repo and then copy the result into mythtv, or copy the ffmpeg master clean into mythtv and apply all of our customizations.

Are these changes in the pull request for the version we have in MythTV master now? If so they will need to be either backported to our copy of the FFmpeg repo or they need to be added to a list of customizations to be done when we copy a new version of FFmpeg to MythTV. (depending which approach we take in future.

My feeling is we should merge or cherry-pick all of the ffmpeg master changes into our ffmpeg master, I don't know whether rebasing would be possible. If a strange bug occurs and we have to bisect, it is most likely to be caused by an FFmpeg commit, not a MythTV commit, so having all the ffmpeg commits in our repo would be useful.

ulmus-scott commented:
I rebased onto https://github.com/MythTV/mythtv/commit/976296fc365f552d1f95cc4c14b7c1ecf6756b2a and force pushed.  So this is on the current master (ignoring your commit to remove the file).

The first 11 commits are primarily changes to MythTV, which enable removing some of the ffmpeg customizations.  (All tested OK with respect to skipping.)

The rest of the changes are copied from https://github.com/ulmus-scott/FFmpeg/commits/rebase/4.4m2

ulmus-scott commented:
Apparently, replacing the deprecated av_free_packet was not as straightforward as I thought.

After reverting that commit, there are no issues skipping.

ulmus-scott commented:
https://github.com/MythTV/mythtv/pull/416/commits/b2d317188362849538c18c315f9e41274a62a49b is correct however (look at context below the change).

However, that change appears to be from https://github.com/MythTV/mythtv/commit/51c59da859f9e9b2a3f6a0a9ec0ede1ee5be393f#diff-4b6a432208acd44651f07cdde943169aba8c5611b26a83764ad784d00c250f7fR1727 in 2005, so that whole added hunk can probably be removed.  What's your opinion?

Also, let me know if you want me to rebase to remove the reverted commit.

ulmus-scott commented:
Rebased.  Dropped problematic commit and its reversion.  Reworded commit message: https://github.com/MythTV/mythtv/pull/416/commits/4c32f80383de8f1c5cf7214fa6b7024a92916485

Also, added two new commits.

ulmus-scott commented on mythtv/libs/libmythtv/decoders/avformatdecoder.cpp:
```diff
@@ -1312,53 +1312,42 @@ int AvFormatDecoder::OpenFile(MythMediaBuffer *Buffer, bool novideo,
+    // last resort, default to NTSC
     rates.emplace_back(30000.0 / 1001.0);
 
+    auto valid_fps = [](double rate) { return rate < 3.0 || rate > 121.0; };
```
This should probably be called **in**valid_fps

ulmus-scott commented:
libavformat/url.h is an internal FFmpeg header.  It was deprecated a decade ago (when it was part of avio.h).  Functions in avio.h may replace url.h?

ulmus-scott commented:
Note: for testing most of the freesurround changes, `./configure --disable-fftw3`

Final, pending review.

ulmus-scott commented:
TODO:

- [ ] cherry-pick from https://github.com/MythTV/mythtv/pull/456
- [ ] rebase onto master (I will squash some of the fixup commits)

ulmus-scott commented:
The freesurround changes should be standalone, so I'll probably put them in a separate pull request.

ulmus-scott commented:
The freesuround changes are now at https://github.com/MythTV/mythtv/pull/462

bennettpeter commented:
FFmpeg 5.0 has been released. What is your plan with all of these changes? 

The normal process is that I merge the new FFmpeg master and new branch into the MythTV/FFmpeg repository. That sometimes requires a lot of fixing of conflicts that arise in the merge. I may have to merge or cherry-pick commits individually rather than merging the branch, since there have been major changes in FFmpeg.

After successfully merging into our MythTV/FFmpeg repository, I copy the entire contents from MythTV/FFmpeg to the external/FFmpeg directory in MythTV/mythtv. This would wipe out any changes made to mythtv/external/FFmpeg, and likely wipe out most of the changes in here.

I have a process whereby I do not make changes in MythTV/external/FFmpeg without also making them in MythTV/FFmpeg first. Perhaps your pull request should be against MythTV/FFmpeg. Or do you have another process for integrating new versions of FFmpeg?

ulmus-scott commented:
> FFmpeg 5.0 has been released. What is your plan with all of these changes?

I have a work in progress rebase onto 5.0, but mpegts-mythtv.c does not compile.  Also, I skipped https://github.com/ulmus-scott/FFmpeg/commit/b642a799b5b6d35a46038f84ef6c54593e50d485 at the time because those function were moved to demux.c.

All of the changes to FFmpeg should be in one of my branches (probably https://github.com/ulmus-scott/FFmpeg/commits/rebase/4.4m3 ).  I'll see about creating a new branch that has all the changes I have made (or reuse rebase/4.4).

> The normal process is that I merge the new FFmpeg master and new branch into the MythTV/FFmpeg repository. That sometimes requires a lot of fixing of conflicts that arise in the merge. I may have to merge or cherry-pick commits individually rather than merging the branch, since there have been major changes in FFmpeg.
> 
> After successfully merging into our MythTV/FFmpeg repository, I copy the entire contents from MythTV/FFmpeg to the external/FFmpeg directory in MythTV/mythtv. This would wipe out any changes made to mythtv/external/FFmpeg, and likely wipe out most of the changes in here.

The point to making the FFmpeg changes in this PR is so they can be easily tested with MythTV and bisected if necessary.
 
> I have a process whereby I do not make changes in MythTV/external/FFmpeg without also making them in MythTV/FFmpeg first. Perhaps your pull request should be against MythTV/FFmpeg. Or do you have another process for integrating new versions of FFmpeg?

The last few commits are intended to make rebasing easier (when squashed into the other commits already rebased onto FFmpeg/release/4.4).

Regarding rebase instead of merge: https://www.git-scm.com/book/en/v2/Git-Tools-Rerere

My plan is:

1. to reduce our downstream changes in this PR (and silence the warnings when compiling FFmpeg that our modifications added)
2. convert mpegts-mythtv.c to the new packet API in a separate PR.
3. Once those changes are settled:
  1. Replace external/FFmpeg with a pristine, unmodified FFmpeg 5.0 (MythTV will probably not compile, however.)
  2. Add the rebased MythTV changes onto FFmpeg

https://github.com/MythTV/mythtv/blob/e6307f599c940966ea28795465d9e3e7b8065635/mythtv/libs/libmythtv/decoders/avformatdecoder.cpp#L2761 is the only use of av_remove_stream() that MythTV has kept alive.  I'm not sure it is necessary.

I also intend to attempt to upstream some of MythTV's changes to FFmpeg.  I am currently waiting to join the ffmpeg-devel mailing list.

bennettpeter commented:
I have merged this pull request into branch devel/ffmpeg-resync

ulmus-scott commented:
New commits:
* https://github.com/MythTV/mythtv/pull/416/commits/48fc6948413eb692d0b19c955a31316b214ba3d8 technically independent
* remove final libavformat/utils.c warning
  * https://github.com/MythTV/mythtv/pull/416/commits/f082214364d34fc81908b613d2d40b64d122720d
  * https://github.com/MythTV/mythtv/pull/416/commits/25b7583221bc49f66d4b82774a2ce99c4aadf5e1

bennettpeter commented:
Please add the ffmpeg commits to a pull request in the MythTV/FFmpeg repository.

bennettpeter commented:
I have incorporated three new commits into branch devel/ffmpeg-resync

ulmus-scott commented:
@bennettpeter Are we waiting for more testing or until the release issues die down?

I would also like to submit the working commits from my harmonization after this is merged, since those remove the use of the deprecated AVPacket API in mpegts-mythtv.c.  Then you can merge in FFmpeg 5.0, which removed one of those functions.

bennettpeter commented:
We can include your harmonization changes into the ffmpeg-resync branch on top of what is already there now, as well as into the ffmpeg repository master branch, if that makes sense to you. Add those commits either here or to a separate pull request, and to a pull request for the MythTV/FFmpeg repository. We can test that for a while and if it looks good merge the main FFmpeg up to 5.0.

ulmus-scott commented:
Rebased cleanly onto master because of https://github.com/MythTV/mythtv/commit/8d08c75902fe8f6050f1df8b774f6c9e538e8ace , which increased the DB schema version.

New commits start at [mythavutil.h: create MythAVBufferRef](https://github.com/MythTV/mythtv/pull/416/commits/9e4634fb42ab58ccef475755cc6d91d5772c2ee8).  This should be 39 new commits for harmonizing mpegts-mythtv.

bennettpeter commented:
These changes affect many parts of the system. I do not understand the ramifications of all of this. The aim of simplifying FFmpeg changes is good, but I do not have the understanding of FFmpeg internals to know whether all of this is valid. I cannot proceed with this.

ulmus-scott commented:
@bennettpeter Should I split the mpegts harmonization commits into a separate PR so the MythTV changes can be merged?  Or does your comment also apply to the original 70+ commits?

ulmus-scott commented:
I have removed the harmonize commits as that is still a WIP and should be reviewed separately.

ulmus-scott commented:
Rebased onto master without conflict, fixed a typo in a commit message, squashed fixup QtTest commit; final pending review and testing.

ulmus-scott commented:
I have rebased this onto master and made the following changes:

- renamed ff_codec_type_string to AVMediaTypeToString in mythtv/libs/libmythtv/decoders/avformatdecoder.cpp
- dropped changes concerning the use of FFmpeg's internal header libavformat/url.h since it only includes public headers and it is more involved to replace its use and I have not fully investigated it.
- only change to ffmpeg:
```
git diff ffmpeg-cleanup-backup -- mythtv/external/FFmpeg
diff --git a/mythtv/external/FFmpeg/libavcodec/libavcodec.v b/mythtv/external/FFmpeg/libavcodec/libavcodec.v
index 3f3dc96da0..f1d5e5bc77 100644
--- a/mythtv/external/FFmpeg/libavcodec/libavcodec.v
+++ b/mythtv/external/FFmpeg/libavcodec/libavcodec.v
@@ -1,11 +1,6 @@
 LIBAVCODEC_MAJOR {
     global:
         av*;
-        ff_ue_golomb_vlc_code;
-        ff_golomb_vlc_len;
-        ff_se_golomb_vlc_code;
-        ff_fft_init;
-        ff_fft_end;
     local:
         *;
 };
```

ulmus-scott commented:
I have now fixed a write past the end bug that I accidentlally introduced by not removing https://github.com/MythTV/mythtv/pull/416/commits/f49697bd106b21fb5b9446c9642bcbcf44e20a3e#diff-5db8e409da1f0c85745590bb76ec9b66bd841819f560424b6056d8ea35d6cdcbL222 when changing https://github.com/MythTV/mythtv/pull/416/commits/f49697bd106b21fb5b9446c9642bcbcf44e20a3e#diff-5db8e409da1f0c85745590bb76ec9b66bd841819f560424b6056d8ea35d6cdcbR137 .

It now tests fine with all of my samples.

bennettpeter commented:
I have reset devel/ffmpeg and merged the pull request again.

bennettpeter commented:
closed by mistake

bennettpeter commented:
I have rebased devel/ffmpeg-resync on to master, and resolved the avformatdecoder conflict

ulmus-scott commented:
> I am testing the devel/ffmpeg-resync branch.
> 
> This is an anomaly with the duration of ts files in the videos directory and recordings without a seek table. The duration is incorrect in the devel/ffmpeg-resync branch as well as in the master branch, but the error is worse in devel/ffmpeg-resync.
> 
> I don't know why there is a difference, maybe duration is just some random number?
> 
> **recording 1 without seek table** mythtv shows length as 0 and cannot skip, current master shows 10 hrs and can skip, actually is 59 min.
> 
> **recording 2 in videos directory without seek table** mythtv shows length as 3346 hours master shows 1hr 40min. actually should be 1hr.
> 
> **recording 2 in videos directory without seek table** shows as 147 hours, master shows 4 minutes actually is 1 min.

Could the problem be https://github.com/MythTV/mythtv/commit/4a278390f3e0c38d4dbe5c3659f457c5f174440f#diff-7d5a046ff4d6a2f59986ccb6fc8dff7f32bdef90d0b0b4231d2327791f0ffd29L1120-L1123 ?

Or maybe this hack: https://github.com/MythTV/mythtv/commit/4eba49f884c4b05dcb6fb14b11d72e6b6437b21e ?  Try removing  that function and see if it is better.  Can you share your samples with me so I can test?  I didn't notice weird durations with my samples.

bennettpeter commented:
vlc gets the time correct and so does exoplayer, which is used in leanfront. Exoplayer does it by reading at start of file to find the first timestamp, the end of the file to find the last timestamp, and subtract the two. It has a number of bytes to search parameter and starts at that number of bytes before the end of file when looking for the last timestamp.. I don't know if this helps or how ffmpeg is calculating the duration.

bennettpeter commented:
Regarding the samples:
The worst problem is the one that shows length as 0 and will not skip. That is 7 GB.
The one that shows duration as 3346 hours is SD and is 849 MB
The one that shows 147 hours is 52.4 MB.  I posted that here --> https://drive.google.com/file/d/1LmkOFUPzNdy7gBJSojKGKcgdJfp9u108/view?usp=sharing

bennettpeter commented:
av_update_stream_timings_video is only called from one place.  I commented out the code which calls it. That made no difference to the error, the short clip still shows 147 hours as duration.

ulmus-scott commented:
@bennettpeter I downloaded your short sample and put it in my Videos directory and scanned for new videos and it all looks good to me when I played it.  Where exactly are you seeing the duration of 147 hours?

bennettpeter commented:
The video is only about 1 minute long. If you start play from beginning then press I it displays the progress graph with the total duration and the current position. That shows 147 hours as the total duration and the progress bar shows nothing since it has only played a few seconds out of the supposed 147 hours.

ulmus-scott commented:
I can't replicate your issue.  For me, whether I bring up the progress bar with `I` or `space` to pause and play, it shows the correct duration "0:xx of 0:41".

ulmus-scott commented:
Could it be related to the theme?  I've tested with both MythCenter-wide and Mythbuntu.

bennettpeter commented:
That is interesting. I am using Steppes, but I don't think it is theme related. I tried mythcenter-wide and it is the same.
Pause shows
0:00:05 of 147:51:12 (147:51:07)
I tried video playback NVDEC, OpenGL normal, VDPAU, all show the same.
Did you do anything different with the file? I just put it in the videos directory and ran mythutil --scanvideos
I will see if I can find out where that duration figure is coming from.

ulmus-scott commented:
> That is interesting. I am using Steppes, but I don't think it is theme related.

I just tested Steppes and I still see 0:41, so probably not theme related then.

> Did you do anything different with the file? I just put it in the videos directory and ran mythutil --scanvideos

I put the file in my videos directory and then scanned the videos directory via the frontend: Media Library -> Watch Videos -> `M` -> Scan for Changes

bennettpeter commented:
I traced the calculation of the duration in the code, for this particular file.

This is what happens:

avformatdecoder.cpp:1117
	dur = duration_cast<std::chrono::seconds>(m_playbackInfo->QueryTotalDuration());
dur is 0
avformatdecoder.cpp:1122
	dur = duration_cast<std::chrono::seconds>(av_duration(m_ic->duration));
m_ic is a AVFormatContext
AVFormatContext has a huge negative value for duration
m_ic->duration is -9223372036854775808
dur is -9223372036854
avformatdecoder.cpp:1143
It recalculates because the dur was not greater than 0
m_bitrate is 1576
bytespersec is 98.5
This is off by a factor of 1000 because it fails to take into account that the bitrate is in kb/sec not bits/sec.
Also this does an extra division by 2 because of some error with pvr-250. What if this was not recorded on a pvr-250?
The bitrate was an estimate, so whether it was recorded on a pvr or not is irrelevant.
So it is also off by another factor of 2.
The actual bitrate from mediainfo is 8981 kb/s, i.e. 8981000 bits/sec, from ffprobe is 9868 kb/s or 9868000
actual file size is (52,428,800 bytes) so it calculates duration as 52,428,800 / 98.5 = 532272 seconds, which is 147.85 hours.

Incorrect bitrate
How does it get that value of 1576?
avformatdecoder.cpp:1920
It goes through each stream accumulation all the bitrates reported by ffmpeg.
For the video bitrate, at avformatdecoder.cpp:1951 it finds zero reported, so it uses an estimate, which comes to 1000000.
Total bitrate after going through all streams is 
at avformatdecoder.cpp:2030
m_bitrate = 1576000
then at avformatdecoder.cpp:2398 it divides it by 1000
m_bitrate = 1576

ffprobe does get the correct duration and bitrates
  Duration: 00:00:42.50, start: 27505.376500, bitrate: 9868 kb/s

Perhaps you can trace it and see what it is doing different on your system.

ulmus-scott commented:
`mythfrontend --logpath=./log --verbose libav:debug,playback:info` is probably a better verbose option so the ScanStreams() logs are printed.
[mythfrontend.20220627205653.15617.log](https://github.com/MythTV/mythtv/files/8995600/mythfrontend.20220627205653.15617.log)

bennettpeter commented:
I applied your patch to my build of the devel/ffmpeg-resync branch. Below is the log you asked for, which confirms that m_ic->duration is returning that large negative number. Strange that there are different results for you and me with the same code and the same recording.

[mythfrontend.20220629125717.66376.log](https://github.com/MythTV/mythtv/files/9010691/mythfrontend.20220629125717.66376.log)

ulmus-scott commented:
Your problem is:
```
2022-06-29 08:57:44.769953 I [66376/66376] CoreContext avformatdecoder.cpp:951 (OpenFile) - AFD: Using FFmpeg MPEG-TS demuxer (forced)
```

You get `Probe buffer size limit of 5000000 bytes reached`:
```
2022-06-29 08:57:44.940965 D [66376/66376] CoreContext avformatdecoder.cpp:243 (myth_av_log) - [mpegts-ffmpeg @ 00007f7dffa47160] Probe buffer size limit of 5000000 bytes reached
2022-06-29 08:57:44.940991 D [66376/66376] CoreContext avformatdecoder.cpp:243 (myth_av_log) - [mpegts-ffmpeg @ 00007f7dffa47160] stream 0: start_time: 27506.1 duration: NOPTS
2022-06-29 08:57:44.940994 D [66376/66376] CoreContext avformatdecoder.cpp:243 (myth_av_log) - [mpegts-ffmpeg @ 00007f7dffa47160] stream 1: start_time: 27505.5 duration: NOPTS
2022-06-29 08:57:44.940996 D [66376/66376] CoreContext avformatdecoder.cpp:243 (myth_av_log) - [mpegts-ffmpeg @ 00007f7dffa47160] stream 2: start_time: 27505.4 duration: NOPTS
2022-06-29 08:57:44.940998 D [66376/66376] CoreContext avformatdecoder.cpp:243 (myth_av_log) - [mpegts-ffmpeg @ 00007f7dffa47160] stream 3: start_time: 27507.1 duration: NOPTS
2022-06-29 08:57:44.941000 D [66376/66376] CoreContext avformatdecoder.cpp:243 (myth_av_log) - [mpegts-ffmpeg @ 00007f7dffa47160] stream 4: start_time: 27505.5 duration: NOPTS
2022-06-29 08:57:44.941001 D [66376/66376] CoreContext avformatdecoder.cpp:243 (myth_av_log) - [mpegts-ffmpeg @ 00007f7dffa47160] stream 5: start_time: NOPTS duration: NOPTS
2022-06-29 08:57:44.941003 D [66376/66376] CoreContext avformatdecoder.cpp:243 (myth_av_log) - [mpegts-ffmpeg @ 00007f7dffa47160] stream 6: start_time: NOPTS duration: NOPTS
2022-06-29 08:57:44.941004 D [66376/66376] CoreContext avformatdecoder.cpp:243 (myth_av_log) - [mpegts-ffmpeg @ 00007f7dffa47160] stream 7: start_time: NOPTS duration: NOPTS
2022-06-29 08:57:44.941006 D [66376/66376] CoreContext avformatdecoder.cpp:243 (myth_av_log) - [mpegts-ffmpeg @ 00007f7dffa47160] stream 8: start_time: NOPTS duration: NOPTS
2022-06-29 08:57:44.941007 D [66376/66376] CoreContext avformatdecoder.cpp:243 (myth_av_log) - [mpegts-ffmpeg @ 00007f7dffa47160] stream 9: start_time: NOPTS duration: NOPTS
2022-06-29 08:57:44.941009 D [66376/66376] CoreContext avformatdecoder.cpp:243 (myth_av_log) - [mpegts-ffmpeg @ 00007f7dffa47160] stream 10: start_time: NOPTS duration: NOPTS
2022-06-29 08:57:44.941012 D [66376/66376] CoreContext avformatdecoder.cpp:243 (myth_av_log) - [mpegts-ffmpeg @ 00007f7dffa47160] format: start_time: 27505.4 duration: NOPTS (estimate from bit rate) bitrate=0 kb/s
2022-06-29 08:57:44.941026 W [66376/66376] CoreContext avformatdecoder.cpp:243 (myth_av_log) - [mpegts-ffmpeg @ 00007f7dffa47160] Could not find codec parameters for stream 5 (Unknown: none ([5][0][0][0] / 0x0005)): unknown codec
Consider increasing the value for the 'analyzeduration' (60000000) and 'probesize' (5000000) options
2022-06-29 08:57:44.941029 W [66376/66376] CoreContext avformatdecoder.cpp:243 (myth_av_log) - [mpegts-ffmpeg @ 00007f7dffa47160] Could not find codec parameters for stream 6 (Unknown: none ([5][0][0][0] / 0x0005)): unknown codec
Consider increasing the value for the 'analyzeduration' (60000000) and 'probesize' (5000000) options
2022-06-29 08:57:44.941032 W [66376/66376] CoreContext avformatdecoder.cpp:243 (myth_av_log) - [mpegts-ffmpeg @ 00007f7dffa47160] Could not find codec parameters for stream 7 (Unknown: none ([5][0][0][0] / 0x0005)): unknown codec
Consider increasing the value for the 'analyzeduration' (60000000) and 'probesize' (5000000) options
2022-06-29 08:57:44.941035 W [66376/66376] CoreContext avformatdecoder.cpp:243 (myth_av_log) - [mpegts-ffmpeg @ 00007f7dffa47160] Could not find codec parameters for stream 8 (Unknown: none ([5][0][0][0] / 0x0005)): unknown codec
Consider increasing the value for the 'analyzeduration' (60000000) and 'probesize' (5000000) options
2022-06-29 08:57:44.941037 W [66376/66376] CoreContext avformatdecoder.cpp:243 (myth_av_log) - [mpegts-ffmpeg @ 00007f7dffa47160] Could not find codec parameters for stream 9 (Unknown: none ([5][0][0][0] / 0x0005)): unknown codec
Consider increasing the value for the 'analyzeduration' (60000000) and 'probesize' (5000000) options
2022-06-29 08:57:44.941040 W [66376/66376] CoreContext avformatdecoder.cpp:243 (myth_av_log) - [mpegts-ffmpeg @ 00007f7dffa47160] Could not find codec parameters for stream 10 (Unknown: none ([5][0][0][0] / 0x0005)): unknown codec
Consider increasing the value for the 'analyzeduration' (60000000) and 'probesize' (5000000) options
2022-06-29 08:57:44.941141 D [66376/66376] CoreContext avformatdecoder.cpp:243 (myth_av_log) - [mpegts-ffmpeg @ 00007f7dffa47160] After avformat_find_stream_info() pos: 5465348 bytes read:5472256 seeks:0 frames:727
2022-06-29 08:57:44.941159 I [66376/66376] CoreContext avformatdecoder.cpp:1917 (ScanStreams) - AFD: Stream #0: ID: 0x9
```

instead of `All info found`:
```
2022-06-27 16:56:59.616345 D [15617/15617] CoreContext avformatdecoder.cpp:243 (myth_av_log) - [mpegts @ 00007f59b3e51160] All info found
...
2022-06-27 16:56:59.618008 D [15617/15617] CoreContext avformatdecoder.cpp:243 (myth_av_log) - [mpegts @ 00007f59b3e51160] stream 0: start_time: 27506.1 duration: 41.74
2022-06-27 16:56:59.618011 D [15617/15617] CoreContext avformatdecoder.cpp:243 (myth_av_log) - [mpegts @ 00007f59b3e51160] stream 1: start_time: 27505.5 duration: 41.544
2022-06-27 16:56:59.618013 D [15617/15617] CoreContext avformatdecoder.cpp:243 (myth_av_log) - [mpegts @ 00007f59b3e51160] stream 2: start_time: 27505.4 duration: 41.64
2022-06-27 16:56:59.618015 D [15617/15617] CoreContext avformatdecoder.cpp:243 (myth_av_log) - [mpegts @ 00007f59b3e51160] stream 3: start_time: 27507.1 duration: NOPTS
2022-06-27 16:56:59.618017 D [15617/15617] CoreContext avformatdecoder.cpp:243 (myth_av_log) - [mpegts @ 00007f59b3e51160] stream 4: start_time: 27505.5 duration: 41.472
2022-06-27 16:56:59.618020 D [15617/15617] CoreContext avformatdecoder.cpp:243 (myth_av_log) - [mpegts @ 00007f59b3e51160] format: start_time: 27505.4 duration: 42.5 (estimate from pts) bitrate=9868 kb/s
2022-06-27 16:56:59.618107 D [15617/15617] CoreContext avformatdecoder.cpp:243 (myth_av_log) - [mpegts @ 00007f59b3e51160] After avformat_find_stream_info() pos: 0 bytes read:1429648 seeks:2 frames:141
```

Your duration is `duration: NOPTS (estimate from bit rate) bitrate=0 kb/s` while mine is `duration: 42.5 (estimate from pts) bitrate=9868 kb/s`.

The problem is
https://github.com/MythTV/mythtv/blob/b987f5bbaff2d820bd1967348a5abb98bcb82dd2/mythtv/external/FFmpeg/libavformat/utils.c#L2973-L2978 also needs `|| !strcmp(ic->iformat->name, "mpegts-ffmpeg")`.

ulmus-scott commented:
New commit here and https://github.com/MythTV/FFmpeg/pull/5 to fix your issue.  By forcing the ffmpeg demuxer I could replicate your issue and the new commit fixes it.

bennettpeter commented:
I added that commit, now it all looks good.

ulmus-scott commented:
@bennettpeter Should we issue a new call for testing to mythtv-dev for this set of changes?

Also, should I close the other pull requests that I had split this into?

bennettpeter commented:
> @bennettpeter Should we issue a new call for testing to mythtv-dev for this set of changes?

Sure, do you want to do it or shall I ?
> Also, should I close the other pull requests that I had split this into?

Yes

ulmus-scott commented:
> > @bennettpeter Should we issue a new call for testing to mythtv-dev for this set of changes?
> 
> Sure, do you want to do it or shall I ?
> 
> > Also, should I close the other pull requests that I had split this into?
> 
> Yes

Done and done.

ulmus-scott commented:
@bennettpeter new commit fixes the LiveTV issue Piotr was having.  I can rebase and remove that commit and its reversion if you want.

bennettpeter commented:
I have rebased the devel/ffmpeg-resync branch. Are we ready to merge it into master? If so, I will check with the developers and if nobody has an issue I can merge it.

ulmus-scott commented:
The only issues reported to the mailing list were by Piotr and I fixed the LiveTV one.  I'm not sure if the other VAAPI issue he briefly mentioned is new or not.  His .mkv sample issue also effected master so was not new.

I had hoped for affirmative "Works fine for me" replies to my call for testing, but otherwise I think it is ready to merge.

ulmus-scott commented:
I have rebased onto master, fixing the revert commit to add back https://github.com/MythTV/mythtv/blob/fb9457e9aa4de597c2bcd17aec3bfbe6f6da1bbf/mythtv/libs/libmythtv/decoders/avformatdecoder.cpp#L203 and fix the referenced commit hash (which may get clobbered again, *shrug*).

bennettpeter commented:
I prefer to merge what we currently have in devel/ffmpeg-resync. I will rebase it again before merging.

bennettpeter commented:
This has now been merged into master
